item.lua

```diff
@@ -0,0 +1,294 @@
+        tab,
+        function(item) return dfhack.items.getDescription(item, 1) == desc end
+    )
+end
+
+function item.condition_material(tab, material)
+    addCondition(
+        tab,
+        function(item) return dfhack.matinfo.decode(item):toString() == material end
+    )
+end
+
+function item.condition_wear(tab, level, upper)
+    if upper then
+        addCondition(tab, function(item) return item.wear <= level end)
+    else
+        addCondition(tab, function(item) return item.wear >= level end)
+    end
+end
+
+function item.condition_quality(tab, levelst, upper)
+    local level = argparse.nonnegativeInt(levelst, 'wear')
+    if upper then
+        addCondition(tab, function(item) return item.quality <= level end)
+    else
+        addCondition(tab, function(item) return item.quality >= level end)
+    end
+end
+
+function item.act(action,conditons,options)
+    local count = 0
+    local changed = 0
+    local invisible = 0
+    for _, item in pairs(df.global.world.items.other.IN_PLAY) do
+        -- never act on items used for constructions/building materials and carried by hostiles
+        -- also skip artifacts, unless explicitly told to include them
+        if  item.flags.construction or
+            item.flags.in_building or
+            item.flags.hostile or
+            (item.flags.artifact and not options.artifact)
+        then
+            goto skipitem
+        end
+
+        -- check conditions provided via options
+        for _, condition in pairs(conditons) do
+            if not condition(item) then goto skipitem end
+        end
+
+        -- only try to melt things that are actually meltable
+        -- TOTHINK: treated as implicit filter, should we count the item?
+        if action == 'melt' and not isMeltable(item) then
+            goto skipitem
+        end
+
+        -- item matches the filters
+        count = count + 1
+
+        -- skip items that are in unrevealed parts of the map
+        local x,y,z = dfhack.items.getPosition(item)
+        if x and not dfhack.maps.isTileVisible(x,y,z) then
+            invisible = invisible+1
+            goto skipitem
+        end
+
+        -- carry out the action
+        if action == 'forbid' and not item.flags.forbid then
+            item.flags.forbid = true
+            changed = changed + 1
+        elseif action == 'unforbid' and item.flags.forbid then
+            item.flags.forbid = false
+            changed = changed + 1
+        elseif action == 'dump' and not item.flags.dump then
+            item.flags.dump = true
+            changed = changed + 1
+        elseif action == 'undump' and item.flags.dump then
+            item.flags.dump = false
+            changed = changed + 1
+        elseif action == 'melt' and not item.flags.melt then
+            item.flags.melt = true
+            changed = changed + 1
+        elseif action == 'unmelt' and item.flags.melt then
+            item.flags.melt = false
+            changed = changed + 1
```

Comment:
I wasn't aware that there is double bookkeeping required for melting. In this case, maybe the melting logic should be encapsulated somewhere, potentially somewhere more basic than `logistics`.

Reply:
@myk002 I took a second look at this, and I don't even fully understand what is required here. I'm inclined to just hold back on melting until there is an API for melting that can be called safely. Maybe this logic can be moved to (what is exported as) `dfhack.items`? There is already some trade logic there.

Reply:
it makes sense to expose that as a library function call; both `canMelt()` and `setMelt(bool)`

Reply:
@myk002 Would you be willing to add those two to the Steam `testing` release. Then I will rework the PR to make use of them.

Reply:
I will, but after -r4 gets tagged later this week. We need to make a quick release to fix some current impactful bugs.

Reply:
@myk002 Can we make progress on this?

Reply:
@myk002 Just to avoid any potential duplication of work: I have started to work on the C++ side of this. I should be able to create a PR before Wednesday.